api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+
+	var errs []error
+
+	// calculate for scaling up
+	resourceProviders, scaleUpErrs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+
+	errs = append(errs, scaleUpErrs...)
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+
+	// do the scaling
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	if err := e.scaleTo(clusterName, desiredScale, unusedProviders); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.MultiError(errs)
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*ResourceProvider, error) {
+	env, err := e.EnvironmentProvider.Read(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &ResourceProvider{}
+	resource.AvailableMemory = InstanceSizes[env.InstanceSize]
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) calculateScaleDown(clusterName string, resourceProviders []*ResourceProvider) []*ResourceProvider {
+	var unusedProviders []*ResourceProvider
+
+	for _, provider := range resourceProviders {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	return unusedProviders
+}
+
+func (e *EnvironmentScaler) calculateScaleUp(clusterName string, resourceProviders []*ResourceProvider, resourceConsumers []ResourceConsumer) ([]*ResourceProvider, []error) {
+	sorts := []struct {
+		Errs      []error
+		Providers []*ResourceProvider
+		SortBy    string
+	}{
+		{[]error{}, resourceProviders, "cpu"},
+		{[]error{}, resourceProviders, "mem"},
+	}
+
+	for _, s := range sorts {
+		switch s.SortBy {
+		case "cpu":
+			sortProvidersByCPU(s.Providers)
+			sortConsumersByCPU(resourceConsumers)
+
+		case "mem":
+			sortProvidersByMemory(s.Providers)
+			sortConsumersByMemory(resourceConsumers)
+		}
+
+		sortProvidersByUsage(s.Providers)
+
+		for _, consumer := range resourceConsumers {
+			hasRoom := false
+
+			for _, provider := range s.Providers {
+				if provider.HasResourcesFor(consumer) {
+					hasRoom = true
+					provider.SubtractResourcesFor(consumer)
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment."
+					err := fmt.Errorf(text)
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				newProvider.SubtractResourcesFor(consumer)
+				s.Providers = append(s.Providers, newProvider)
+			}
+		}
+	}
+
+	if len(sorts[0].Providers) > len(sorts[1].Providers) {
```

Comment:
This bit of code is a bit mysterious to me

Reply:
@diemonster We do two different allocations-of-consumers-among-providers. In one scenario, we sort everything by CPU and then allocate consumers among providers, adding new providers as necessary. In another scenario, we sort everything by memory and then allocate, adding new providers as necessary.

This line is for choosing the allocation which creates the fewest new resources. In other words, if we create two new providers when prioritizing compactness by CPU but only one new provider when prioritizing compactness by memory, we want to move forward with that second array of providers because it requires the creation of fewer new resources.

Reply:
+1 (It's choosing the more optimal solution, but it's not very readable in its current design)

Reply:
I've got another idea that might be better.